src/object.cr

```diff
@@ -86,72 +86,65 @@ class Object
-  # to be embedded inside other expressions, sometimes providing
-  # more information about this object.
+  # Prints a nicely readable and concise string representation of this object,
+  # typically intended for users, to *io*.
   #
-  # `#inspect` (and `#inspect(io)`) are the methods used when
-  # you invoke `#to_s` or `#inspect` on an object that holds
-  # other objects and wants to show them. For example when you
-  # invoke `Array#to_s`, `#inspect` will be invoked on each element:
+  # This method is called when an object is used in string interpolation:
+  # In `"foo #{bar} baz"`, the value interpolated for `#{foo}` is the result of
+  # `foo.to_s`.
   #
-  # ```
-  # ary = ["one", "two", "three, etc."]
-  # ary.inspect # => ["one", "two", "three, etc."]
-  # ```
+  # Implementations must not append `self` to *io* which would lead to an
+  # endless loop.
   #
-  # Note that if Array invoked `#to_s` on each of the elements
-  # above, the output would have been this:
-  #
-  # ```
-  # ary = ["one", "two", "three, etc."]
-  # # If inspect invoked to_s on each element...
-  # ary.inspect # => [one, two, three, etc.]
-  # ```
-  #
-  # Note that it's not clear how many elements the array has,
-  # or which are they, because `#to_s` doesn't guarantee that
-  # the string representation is clearly delimited (in the case
-  # of `String` the quotes are not shown).
+  # See `#inspect(IO)` for related method.
+  abstract def to_s(io : IO) : Nil
+
+  # Returns an unambiguous and information-rich string representation of this
+  # object, typically intended for developers.
   #
-  # Also note that sometimes the output of `#inspect` will look
-  # like a Crystal expression that will compile, but this isn't
-  # always the case, nor is it necessary. Notably, `Reference#inspect`
-  # and `Struct#inspect` return values that don't compile.
+  # This method should usually **not** be overridden. It delegates to
+  # `#inspect(IO)` which can be overridden for custom implementations.
   #
-  # Classes must usually **not** override this method. Instead,
-  # they must override `inspect(io)`, which must append to the
-  # given `IO` object.
+  # See `#to_s` for related method.
   def inspect : String
     String.build do |io|
       inspect io
     end
   end
 
-  # Appends a string representation of this object
-  # to the given `IO` object.
+  # Prints an unambiguous and information-rich string representation of this
+  # object, typically intended for developers, to *io*.
+  #
+  # The representation should be self-contained with the end being clearly
+  # identifiable.
+  # A Crystal expression that recreates an object with the same value (given an
+  # identical environment) is an ideal representation. It doesn't need to be
+  # actual code that compiles, but a resemblance is recommended.
+  # When this is not possible (for example for non-printable internal state)
+  # another useful description will do.
```

Comment:
Let's be honest, this is a lie. It's true for literal'ish values, but beyond that the representation never even tries represent valid code. The default implementations for `Reference` and `Struct` make a good example of my point here. You wouldn't want the `inspect` representation to be like `SomeStruct.new(a: ..., b: ...)` as that's noisy and a little weird.

 So I'm not sure this is true for people that are not standard library/compiler developers like us. You don't work much with "custom" objects that are literals elsewhere.

Reply:
Maybe the wording could be a bit relaxed, but in general I think it's a good idea to try to print Crystal expressions. That's concise and unambiguous. `Struct#inspect` actually isn't far from Crystal code.
Simple structs/classes which are just containers for a vew ivars are already covered by the respective default implementations. These notes are particularly relevant for more advanced types. For example custom collection types. They can actually use array-like or hash-like literals. `HTTP::Params{"foo" => "bar"}` would be a perfect `inspect` representation. Currently, it's more bulky: `HTTP::Params(@raw_params={"foo" => ["bar"]})`. That's an example from stdlib but you have such types in shards, too.

Reply:
I'm also not entirely happy with this wording, but it's more of style than anything else. Perhaps we can pick a couple of (good) examples from the stdlib instead?

Reply:
I can't add a suggestion to the entire block, so here it goes in plain 🤷 

```
  # Prints to *io* an unambiguous and information-rich string representation of this
  # object, typically intended for developers.
  #
  # It is similar to `#to_s(IO)`, but often provides more information. Ideally, it should
  # contain sufficient information to be able to recreate an object with the same value
  # (given an identical environment).
  #
  # For types that don't provide a custom implementation of this method,
  # default implementation delegates to `#to_s(IO)`. This said, it is advisable to 
  # have an appropriate `#inspect` implementation on every type. Default
  # implementations are provided by `Struct#inspect` and `Reference#inspect`.
  #
  # `::p` and `::p!` use this method to print an object in `STDOUT`.
```